phoenix-core/src/main/java/org/apache/phoenix/coprocessor/IndexRebuildRegionScanner.java

```diff
@@ -0,0 +1,583 @@
+                    scanMaxTs, Bytes.toBytes(errorMsg));
+            hTable.put(put);
+        } catch (IOException e) {
+            exceptionMessage = "LogToIndexToolOutputTable failed " + e;
+        }
+    }
+
+    private long getMaxTimestamp(Result result) {
+        long ts = 0;
+        for (Cell cell : result.rawCells()) {
+            if (ts < cell.getTimestamp()) {
+                ts = cell.getTimestamp();
+            }
+        }
+        return ts;
+    }
+
+    private void verifySingleIndexRow(Result indexRow, final Put dataRow) throws IOException {
+        ValueGetter valueGetter = new SimpleValueGetter(dataRow);
+        long ts = 0;
+        for (List<Cell> cells : dataRow.getFamilyCellMap().values()) {
+            if (cells == null) {
+                break;
+            }
+            for (Cell cell : cells) {
+                if (ts < cell.getTimestamp()) {
+                    ts = cell.getTimestamp();
+                }
+            }
+        }
+        Put indexPut = indexMaintainer.buildUpdateMutation(GenericKeyValueBuilder.INSTANCE,
+                valueGetter, new ImmutableBytesWritable(dataRow.getRow()), ts, null, null);
+        if (indexPut == null) {
+            // This means the index row does not have any covered columns. We just need to check if the index row
+            // has only one cell (which is the empty column cell)
+            if (indexRow.rawCells().length == 1) {
+                return;
+            }
+            String errorMsg = "Expected to find only empty column cell but got "
+                    + indexRow.rawCells().length;
+            exceptionMessage = "Index verify failed - " + errorMsg + indexHTable.getName();
+            logToIndexToolOutputTable(dataRow.getRow(), indexRow.getRow(), ts, getMaxTimestamp(indexRow), errorMsg);
+            throw new IOException(exceptionMessage);
+        }
+        else {
+            // Remove the empty column prepared by Index codec as we need to change its value
+            removeEmptyColumn(indexPut, indexMaintainer.getEmptyKeyValueFamily().copyBytesIfNecessary(),
+                    indexMaintainer.getEmptyKeyValueQualifier());
+        }
+        // Add the empty column
+        indexPut.addColumn(indexMaintainer.getEmptyKeyValueFamily().copyBytesIfNecessary(),
+                indexMaintainer.getEmptyKeyValueQualifier(), ts, VERIFIED_BYTES);
+        int cellCount = 0;
+        for (List<Cell> cells : indexPut.getFamilyCellMap().values()) {
+            if (cells == null) {
+                break;
+            }
+            for (Cell expectedCell : cells) {
+                byte[] family = CellUtil.cloneFamily(expectedCell);
+                byte[] qualifier = CellUtil.cloneQualifier(expectedCell);
+                Cell actualCell = indexRow.getColumnLatestCell(family, qualifier);
+                if (actualCell == null) {
+                    exceptionMessage = "Index verify failed - Missing cell " + indexHTable.getName();
```

Comment:
nit: some extract methods around here would make this easier to read. maybe one function for cell not found and another for checking all the columns?